web/packages/core/src/i18n.ts

```diff
@@ -0,0 +1,123 @@
+import { FluentBundle, FluentResource } from "@fluent/bundle";
+import { negotiateLanguages } from "@fluent/langneg";
+import type { FluentVariable } from "@fluent/bundle";
+
+interface FileBundle {
+    [filename: string]: string;
+}
+
+interface LocaleBundle {
+    [locale: string]: FileBundle;
+}
+
+// This is automatically populated by `tools/bundle_texts.js` via a postbuild script
+const BUNDLED_TEXTS: LocaleBundle = {
+    /* %BUNDLED_TEXTS% */
+};
+
+const bundles: Record<string, FluentBundle> = {};
+
+for (const locale in BUNDLED_TEXTS) {
+    const bundle = new FluentBundle(locale);
+    const files = BUNDLED_TEXTS[locale];
+    if (files) {
+        for (const filename in files) {
+            const text = files[filename];
+            if (text) {
+                for (const error in bundle.addResource(
+                    new FluentResource(text)
+                )) {
+                    console.error(
+                        `Error in text for ${locale} ${filename}: ${error}`
+                    );
+                }
+            }
+        }
+    }
+    bundles[locale] = bundle;
+}
+
+/**
+ * Gets the localised text for the given locale and text ID.
+ *
+ * If the locale does not contain a text for this ID, it will return null.
+ *
+ * @param locale Locale to prefer when retrieving text, ie "en-US"
+ * @param id ID of the text to retrieve
+ * @param args Any arguments to use when creating the localised text
+ * @returns Localised text or null if not found
+ */
+function tryText(
+    locale: string,
+    id: string,
+    args?: Record<string, FluentVariable> | null
+): string | null {
+    const bundle = bundles[locale];
+    if (bundle !== undefined) {
+        const message = bundle.getMessage(id);
+        if (message !== undefined && message.value) {
+            return bundle.formatPattern(message.value, args);
+        }
+    }
+    return null;
+}
+
+/**
+ * Gets the localised text for the given text ID.
+ *
+ * The users preferred locales are used, in priority order, to find the given text.
+ *
+ * If no text is found for any preferred locale, en-US will be used.
+ * If en-US does not contain a text for this ID, an error will be logged and the ID itself will be returned.
+ *
+ * @param id ID of the text to retrieve
+ * @param args Any arguments to use when creating the localised text
+ * @returns Localised text
+ */
+export function text(
+    id: string,
+    args?: Record<string, FluentVariable> | null
+): string {
+    const locales = negotiateLanguages(
```

Comment:
Nit: Could it make sense to place the call to and result of `negotiateLanguages()` elsewhere? Like, instead of calling `negotiateLanguages()` on every call to `text()`, just save the result once and use it there? Though, that depends on whether the arguments to `negotiateLanguages()`, or relevant state of `Fluent`, can change after initialization.

Looking at for instance https://github.com/mozilla/fx-private-relay/blob/eb21d4b31f9f220e8d5d32d233ba4a6b6e728a30/frontend/src/functions/getL10n.ts#L38 (though this is a random file I found through searching on Github, `negotiateLanguages()` feels a bit to me like an initialization call of sorts, though that might not be appropriate in all cases.

On the other hand, while this https://github.com/common-voice/sentence-collector/blob/4b2b77299850b5570617cc8e64c0d11e34ee581d/web/src/l10n.tsx#L58 indicates that the function is rarely called, the localization appears to be able to change during runtime, like if the user changes locale, so as a first approximation, keeping it as it is is presumably fine and more correct than only calculating it once. A more granular method might be to do something similar to that line, where `negotiateLanguages()` is called whenever the locale (or other state?) changes.

Maybe add a TODO comment that this could be made more granular, such as calculating it whenever the locale changes?

But again, this comment is a nitpick, I will leave it up to you.

Reply:
It kinda depends. Do we want to support languages being changed on the fly? Browsers allow that, and `navigator.language`/`navigator.languages` may change at any time.

On the other hand, they *do* fire an event for when it changes. I just didn't think it needed the investment of time to figure that out for the first PR :D